src/librustc_resolve/lifetimes.rs

```diff
@@ -2879,19 +2909,80 @@ fn add_missing_lifetime_specifiers_label(
+        if let Some(generics) = missing_named_lifetime_spots.iter().last() {
+            introduce_suggestion.push(match &generics.params {
+                [] => (generics.span, "<'lifetime>".to_string()),
+                [param, ..] => (param.span.shrink_to_lo(), "'lifetime, ".to_string()),
+            });
+        }
+
+        match (lifetime_names.len(), lifetime_names.iter().next(), snippet) {
+            (1, Some(name), Some("&")) => {
+                err.span_suggestion(
+                    span,
+                    "consider using the named lifetime",
+                    format!("&{} ", name),
+                    Applicability::MaybeIncorrect,
+                );
+            }
+            (1, Some(name), Some("'_")) => {
+                err.span_suggestion(
+                    span,
+                    "consider using the named lifetime",
+                    name.to_string(),
+                    Applicability::MaybeIncorrect,
+                );
+            }
+            (1, Some(name), Some(snippet)) if !snippet.ends_with(">") => {
+                err.span_suggestion(
+                    span,
+                    "consider using the named lifetime",
+                    format!("{}<{}>", snippet, name),
+                    Applicability::MaybeIncorrect,
+                );
+            }
+            (0, _, Some("&")) => {
+                err.span_label(span, "expected named lifetime parameter");
+                if !introduce_suggestion.is_empty() {
+                    introduce_suggestion.push((span, "&'lifetime ".to_string()));
+                    err.multipart_suggestion(
+                        "consider introducing a named lifetime parameter",
+                        introduce_suggestion,
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+            }
+            (0, _, Some("'_")) => {
+                err.span_label(span, "expected named lifetime parameter");
+                if !introduce_suggestion.is_empty() {
+                    introduce_suggestion.push((span, "'lifetime".to_string()));
+                    err.multipart_suggestion(
+                        "consider introducing a named lifetime parameter",
+                        introduce_suggestion,
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+            }
+            (0, _, Some(snippet)) if !snippet.ends_with(">") => {
+                err.span_label(span, "expected named lifetime parameter");
+                if !introduce_suggestion.is_empty() {
+                    introduce_suggestion.push((span, format!("{}<'lifetime>", snippet)));
+                    err.multipart_suggestion(
+                        "consider introducing a named lifetime parameter",
+                        introduce_suggestion,
+                        Applicability::MaybeIncorrect,
+                    );
+                }
+            }
```

Comment:
Also there's a lot of duplication now that I look at this again. I'd introduce a closure:
```rust
let expected_named_param = |span, introduce_suggestion, sugg| {
                err.span_label(span, "expected named lifetime parameter");
                if !introduce_suggestion.is_empty() {
                    introduce_suggestion.push((span, sugg));
                    err.multipart_suggestion(
                        "consider introducing a named lifetime parameter",
                        introduce_suggestion,
                        Applicability::MaybeIncorrect,
                    );
                }
};
```
and then use it in the 3 arms here.